lib/dataSource-recursiveFile.js

```diff
@@ -0,0 +1,166 @@
+                        found.push(currentFullPath);
+                    }
+                    pending--;
+                    (pending === 0) ? callback() : next();
+                }
+            });
+        })();
+    });
+};
+
+/**
+ * Do an asynchronous recursive search of the `rootPath` looking for the `filename`. If the file is found,
+ * the first hit is read and returned via the resolved promise. If no file is found, the promise is rejected.
+ *
+ * @param rootPath {String} the root to search from. This should be  a valid folder.
+ * @param filename {String} filename of the file to find
+ * @param encoding {String} The encoding in which to read the file
+ * @param promise {Promise} A fluid.promise. If an error occurs or no file is found, it will be rejected, else
+ *     it will be resolved with the content of the (first instance of) the found file
+ */
+kettle.dataSource.recursiveFile.readFile = function (rootPath, filename, encoding, promise) {
+    var foundList = [];
+    kettle.dataSource.recursiveFile.findAll(rootPath, filename, foundList, function (err) {
+        if (err) {
+            kettle.dataSource.recursiveFile.rejectPromise(promise, "Error while searching for file... " + err);
+        } else if (foundList.length === 0) {
+            kettle.dataSource.recursiveFile.rejectPromise(promise, "File " + filename + " was not found when searching from path " + rootPath, 404);
+        } else {
+            fs.readFile(foundList[0], encoding, function (err, data) { // read the first file found
+                if (err) {
+                    kettle.dataSource.recursiveFile.rejectPromise("Error while reading file... " + err);
+                } else {
+                    promise.resolve(data);
+                }
+            });
+        }
+    });
+};
+
+/** Central strategy point for all file-system backed DataSource operations (both read and write).
+ * Accumulates options to be sent to the underlying node.js `readFile` or `writeFile` primitives, collects and interprets the
+ * results back into promise resolutions.
+ * @param that {Component} The DataSource itself
+ * @param requestOptions {Object} A merged form of the options sent to the top-level `dataSource.get` method together with relevant
+ * static options configured on the component
+ * @param directModel {Object} The `directModel` argument sent to top-level `dataSource.get/set`
+ */
+kettle.dataSource.recursiveFile.handle = function (that, requestOptions, directModel) {
+    if (!that.options.path) {
+        fluid.fail("Cannot operate recursiveFile dataSource ", that, " without an option named \"path\"");
+    } else if (!that.options.filename) {
+        fluid.fail("Cannot operate recursiveFile dataSource ", that, " without an option named \"filename\"");
+    }
+
+    // resolve path and filename
+    var resolvedPath = kettle.dataSource.URL.resolveUrl(that.options.path, that.options.termMap, directModel, true).replace("//", "/");
```

Comment:
We should take the opportunity to factor datasource-file properly - we can share a front-end which does this resolution, and share a back-end which actually sources the data

Reply:
I'm not sure if this comment is still relevant, since changing the code to do a scan a startup, the path is no longer used and there is generally very little overlap between the two components